### PR TITLE
WT-10362 Implement object_list(), file_exists(), remove(), and rename() in file system class for Azure

### PR DESCRIPTION
**Pull request by lting0804**
- Implement remaining file system functions with python testing 
- Refactored flush finish to call file exists to reduce code repetition
- Rename `file exists` to `file_system_exists`

### CODE REVIEW

File: ext/storage_sources/azure_store/azure_storage_source.cpp [hunk 1]
```diff
@@ -274,41 +258,102 @@ azure_terminate(WT_STORAGE_SOURCE *storage_source, WT_SESSION *session)
...
+{
+    azure_file_system *azure_fs = reinterpret_cast<azure_file_system *>(file_system);
+    std::vector<std::string> objects;
+    std::string complete_prefix;
+
+    *countp = 0;
+
+    if (directory != nullptr) {
+        complete_prefix += directory;
+        // Add a terminating '/' if one doesn't exist.
```
Comment (jiechenbo) [inline ext/storage_sources/azure_store/azure_storage_source.cpp:274]: I think we can improve using cpp notation here, can you investigate into it? @lting0804

File: ext/storage_sources/azure_store/azure_storage_source.cpp [hunk 2]
```diff
@@ -354,9 +412,11 @@ azure_remove(WT_FILE_SYSTEM *file_system, WT_SESSION *session, const char *name,
     WT_UNUSED(name);
     WT_UNUSED(flags);
 
-    return 0;
+    std::cerr << "azure_remove: Object: " << name << ": remove of file not supported." << std::endl;
+    return ENOTSUP;
 }
 
+// POSIX rename, not supported for cloud objects.
```
Comment (korteland) [inline ext/storage_sources/azure_store/azure_storage_source.cpp:419]: Shouldn't we just leave the function pointer in the struct as nullptr? I didn't think this one was mandatory. This possibly goes for remove as well.

File: ext/storage_sources/azure_store/azure_storage_source.cpp [hunk 3]
```diff
@@ -335,17 +380,30 @@ azure_file_system_terminate(WT_FILE_SYSTEM *file_system, WT_SESSION *session)
...
+azure_file_system_exists(
+  WT_FILE_SYSTEM *file_system, WT_SESSION *session, const char *name, bool *existp)
 {
     WT_UNUSED(file_system);
     WT_UNUSED(session);
-    WT_UNUSED(name);
-    WT_UNUSED(existp);
+    azure_file_system *azure_fs = reinterpret_cast<azure_file_system *>(file_system);
 
+    std::cout << "azure_file_system_exists: Checking object: " << name << " exists in Azure."
```
Comment (korteland) [inline ext/storage_sources/azure_store/azure_storage_source.cpp:392]: Looks like some debugging code slipped through, here and at the end of the function.

File: ext/storage_sources/azure_store/azure_storage_source.cpp [hunk 4]
```diff
@@ -335,17 +380,30 @@ azure_file_system_terminate(WT_FILE_SYSTEM *file_system, WT_SESSION *session)
...
+              << std::endl;
+
+    // Check whether the object exists in the cloud.
+    azure_fs->azure_conn->object_exists(name, *existp);
+    if (!*existp) {
```
Comment (korteland) [inline ext/storage_sources/azure_store/azure_storage_source.cpp:397]: What's the motivation behind making this an error? It doesn't seem unreasonable to ask if a file exists and receive "no" as a response. I would've thought that returning non-zero values is for when we fail to check existence at all (e.g. we've lost our network connection).

File: ext/storage_sources/azure_store/azure_storage_source.cpp (hunk 2 above)
Comment (lting0804) [reply ext/storage_sources/azure_store/azure_storage_source.cpp:419]: I might be wrong but is WiredTiger expecting this function to actually be a function?

File: ext/storage_sources/azure_store/azure_storage_source.cpp (hunk 3 above)
Comment (lting0804) [reply ext/storage_sources/azure_store/azure_storage_source.cpp:392]: this is more for explaining the process

File: ext/storage_sources/azure_store/azure_storage_source.cpp (hunk 2 above)
Comment (lting0804) [reply ext/storage_sources/azure_store/azure_storage_source.cpp:419]: @jiechenbo is `remove` and `rename` mandatory

**Review by korteland — APPROVED** [review]
This looks good now - thanks!

**Comment by lting0804** [comment]
i've amended this in https://github.com/wiredtiger/wiredtiger/commit/5dd768eb0d2487367273225a108affead36b3d33 please let me know what your thoughts are

File: ext/storage_sources/azure_store/azure_storage_source.cpp (hunk 1 above)
Comment (jiechenbo) [reply ext/storage_sources/azure_store/azure_storage_source.cpp:285]: Yep, can we use a cpp find function here instead?

**Review by jiechenbo — APPROVED** [review]
LGTM! After comments are addressed

File: ext/storage_sources/azure_store/azure_storage_source.cpp (hunk 1 above)
Comment (lting0804) [reply ext/storage_sources/azure_store/azure_storage_source.cpp:285]: this seems like overengineering if we need to if there's a trailing `/`

File: ext/storage_sources/azure_store/azure_storage_source.cpp [hunk 5]
```diff
@@ -231,24 +227,12 @@ azure_flush_finish(WT_STORAGE_SOURCE *storage_source, WT_SESSION *session,
   WT_FILE_SYSTEM *file_system, const char *source, const char *object, const char *config)
 {
     WT_UNUSED(storage_source);
-    WT_UNUSED(session);
     WT_UNUSED(config);
     WT_UNUSED(source);
-    azure_file_system *azure_fs = reinterpret_cast<azure_file_system *>(file_system);
-
+    bool existsp = false;
```
[the remark this answers is not in the document]
Comment (lting0804) [reply ext/storage_sources/azure_store/azure_storage_source.cpp]: addressed in https://github.com/wiredtiger/wiredtiger/pull/8780/commits/f14bee66c4510fc1a8e079a85c73e78c5a711c6d

File: ext/storage_sources/azure_store/azure_storage_source.cpp (hunk 1 above)
Comment (jiechenbo) [reply ext/storage_sources/azure_store/azure_storage_source.cpp:285]: Hmm I see your point, I'm happy with the way it is!

**Comment by lting0804** [comment]
evergreen merge

**Pull request merged**